### ISSUE TITLE
HasCallStack support for *Note functions

### ISSUE THREAD

**Issue by tebello-thejane** [ndmitchell/safe#18]
Using the `*Note` functions breaks the chain necessary for GHC to produce stack traces. I could annotate all of my functions with `HasCallStack`, but if one of them uses a `*Note` function and that function throws an error, then GHC will produce an unhelpful one line "stack trace".

It would be great if the `safe` library had `HasCallStack =>` annotations (or maybe even `?callStack :: CallStack`, but apparently the call stack mechanism might change in future) on all the `*Note` functions, possibly with CPP used so that these are only used in the supported GHC versions (`base >= 4.9.0.0` for `HasCallStack`, `base >= 4.8.1.0` for the implicit parameter variant).

Does this sound like a good idea? Do you have any qualms with breaking compatibility with other compilers, or using a bunch of CPP in the code? I might give this a whirl if implementing it does not go against your views on the package. :-)

**Comment by minad** [comment]
I also thought about that. It could be done in a very non-intrusive way as I did here https://github.com/minad/intro/blob/master/src/Intro.hs#L895.

**Comment by tebello-thejane** [comment]
There's also the `call-stack` package: https://hackage.haskell.org/package/call-stack

In that time-honored tradition, its documentation is highly lacking, but looking at the source code it appears to at least require the GHC ConstraintKinds extension. @minad's technique would be compatible with any compiler which supports `CPP`.

**Comment by ndmitchell** [comment]
Having the stack trace definitely seems like the right thing to do - it's sad that explicitly taking the time to improve error messages (by using `Note`) makes them worse. Using `ConstraintKinds` makes the haddock output more confusing for users, which makes me inclined to follow the `CPP` route.

I'm not a fan of CPP everywhere, but it does seem the right trade-off here. I'd welcome a pull request.

**Comment by tebello-thejane** [comment]
Eish! So, it just so happens that figuring out which functions might crash, and only adding the call stack annotation on those ones, is a bit... *painful*.

Would marking **everything** with the annotation be overkill? I *think* that most of the functions in `safe` are total, and thus they can't ever crash and print a stack trace...

Also, I was thinking about how the end-users might feel about suddenly having stack traces all over the place, but the GHC team obviously didn't think it would be that bad (and so they added stack traces to `error` and `undefined`)...

**Comment by tebello-thejane** [comment]
Oh, and there's also the fact that [`CPP` seems to explode if it sees a single quote near where it should do a replacement](https://ghc.haskell.org/trac/ghc/ticket/860), so the declaration `Safe.hs#177` is a bit problematic.

Here's my effort thus far: https://github.com/tebello-thejane/safe/commit/74e98c23ae967dd046efe0a8e0964858fb280bed

**Comment by minad** [comment]
I would prefer not to have the call stack on every function. The constraint is a good marker for partial functions.

**Comment by tebello-thejane** [comment]
Here's a workaround for the CPP issue: https://github.com/tebello-thejane/safe/commit/e198ca04451cbe236045fc61e16bfcc84798f01a

Let me go through the code *very* carefully and make sure I haven't missed anything, before I do the PR. Comments welcome, of course. ;-)

**Comment by minad** [comment]
Ah, now I understand what you meant concerning the single quote :)

**Comment by ndmitchell** [comment]
The single quote workaround looks good. Annotating crashing functions with callstack is super useful - it makes me think having a constraint kind named Partial might be a way to go? That way we get information in docs and call stacks. 

It should be easy to see what crashes - basically the note and exact variants - everything else should be safe but I'll review that carefully.

**Comment by minad** [comment]
Is it possible to have a partial constraint kind which is empty? Then we could use this on old ghcs and HasCallStack on newer ones.

~~~ haskell
type Partial = ()
type Partial = HasCallStack
~~~

**Comment by ndmitchell** [comment]
Yep, that was what I was suggesting. Question is if we need to support even older GHC, without constraint kinds, but I don't think so.

**Comment by tebello-thejane** [comment]
[This blog post suggests that `ConstraintKinds` were added to GHC with version 7.4](http://blog.omega-prime.co.uk/?p=127), and so @minad's suggestion would be sufficient to cover all the compilers which `Safe` promises to support (in its `.travis.yml` and `safe.cabal`'s `tested-with` field).

Back to the drawing board it is, then!

**Comment by tebello-thejane** [comment]
@minad: To answer your question about empty constraints, please see the comments to this question: http://stackoverflow.com/q/36369252

In particular, this works:
```Haskell
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE ConstraintKinds #-}
module Main where

import GHC.Stack.Types
import GHC.Exts (Constraint)

main :: IO ()
main = do
  putStrLn "hello world"

type Partial = (() :: Constraint)

f :: Partial => Int -> Int
f = const 5
```

**Comment by minad** [comment]
Great! Then this is what we should use, I guess :)

**Comment by tebello-thejane** [comment]
Please take a gander at this: https://github.com/tebello-thejane/safe/commit/9dd45f2e1dca49a21a66efe669aa1b53938d2d39

In particular:
1. I have put the `Partial` synonym in its own module so it could be re-used.
2. **I have made the `Safe.Partial` module public**, so that users can at least see the documentation for it (I don't expect people to use it themselves).
3. I don't discount the possibility that I missed a function or two; I might have erred on the side of caution (especially with the combinators in `Safe.Util`).

Thoughts (and prayers)?

**Comment by tebello-thejane** [comment]
Alas. I forgot that I needed to enable Travis **before** pushing. I have a feeling that the build will fail due to the edge case of how stack traces worked in GHC 7.10.*.

See here: https://travis-ci.org/tebello-thejane/safe/builds

If necessary, I have have to do another less-than-useful push to trigger the build. But if you have any criticism or suggestions then we can use that to bump the build. :-)

**Comment by tebello-thejane** [comment]
Alright.

I am now **very** certain that the CPP macros cover all the cases in order for this to build on all GHC versions that support constraint synonyms (thanks to `stack`).

Please see commit https://github.com/tebello-thejane/safe/commit/aaf6927d809913a292fc5b7d6ab635c613e70a0d with the bulk of the functionality, and commit https://github.com/tebello-thejane/safe/commit/52c4776a9febe9c46a462b26b6ed30dd6c7ec880 fixing breaks caused by `-Werror`.

Here's the latest (successful) build: https://travis-ci.org/tebello-thejane/safe/builds/194522214

**Comment by ndmitchell** [comment]
I'm happy with pointless commits to trigger Travis - I do them myself all the time :)

It all looks good. Only remark is that I tend to prefer `__GLASGOW_HASKELL__` macros over the `MIN_VERSION_base` as then the code works without Cabal involvement. There is a mapping table at https://github.com/ndmitchell/extra#base-versions.

**Comment by minad** [comment]
@ndmitchell About this __GLASGOW_HASKELL__ macro - is this future proof? It seems to me that MIN_VERSION_base etc is used in more libraries, maybe because it also works for other libraries.

**Comment by tebello-thejane** [comment]
For completeness, here's a commit to change the macro to `__GLASGOW_HASKELL__`: https://github.com/tebello-thejane/safe/commit/3590081940f7244ebd0ef7a614155a6df592fd17 ([Travis build](https://travis-ci.org/tebello-thejane/safe/builds/194715943)).

I don't know why anyone would wanna write/compile Haskell without `Cabal`, but hey... `CPP` always feels dirty/finicky any way you use it, though, as I'm sure we can all agree.

**Comment by ndmitchell** [comment]
@minad it only works with base, which is all safe depends on.

@tebello-thejane I regularly write Haskell and the safe library with GHCi which doesn't set these macros. Where I do use them I make the macro being undefined default to a higher library version - but it's more hassle.

**Comment by ndmitchell** [comment]
Thanks to #19 I think this is all complete. All that remains is to make a fresh release. Let me know if there's anything I've missed.

**Comment by tebello-thejane** [comment]
Hmm... 

I fear I may have made a bit of a booboo. [This library](https://hackage.haskell.org/package/call-stack-0.1.0/docs/Data-CallStack.html) does what we need, and does a slightly better job at it by exporting a function to produce a stack trace at will. 

How do you feel about adding a very small dependency

**Comment by minad** [comment]
@tebello-thejane I don't think so. If you take a look at the source of the call-stack library:

~~~ haskell
{-# LANGUAGE CPP #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE ImplicitParams #-}

module Data.CallStack (
  HasCallStack
, CallStack
, SrcLoc(..)
, callStack
) where

import Data.SrcLoc

#if MIN_VERSION_base(4,8,1)
import qualified GHC.Stack as GHC
#endif

#if MIN_VERSION_base(4,9,0)
import           GHC.Stack (HasCallStack)
#elif MIN_VERSION_base(4,8,1)
type HasCallStack = (?callStack :: GHC.CallStack)
#else
import GHC.Exts (Constraint)
type HasCallStack = (() :: Constraint)
#endif

type CallStack = [(String, SrcLoc)]

callStack :: HasCallStack => CallStack
#if MIN_VERSION_base(4,9,0)
callStack = drop 1 $ GHC.getCallStack GHC.callStack
#elif MIN_VERSION_base(4,8,1)
callStack = drop 2 $ GHC.getCallStack ?callStack
#else
callStack = []
#endif
~~~

It can't do anything more than what we have right now.

**Comment by minad** [comment]
You probably did that already. I just don't think that such tiny libraries make much sense. Consider the nodejs leftpad "library"... :grin:

**Comment by ndmitchell** [comment]
I'm happy with it as it is - no need to worry too much - and hopefully in 10+ years we can simplify things to remove the CPP 😆

**Comment by ndmitchell** [comment]
I've released safe-0.3.12 with this included.
